settings: Fix live `update_page` behavior for user settings.

laurynmm opened the pull request:
Changes `user_settings.update_page` to only update the modified user setting instead of updating all of the user settings on the page. This is modeled on the behavior for updates to the realm user default settings.

@sahil839 - Would you mind looking over this commit for me? This is a follow up on [your feedback on PR 20070 here](https://github.com/zulip/zulip/pull/20070#discussion_r737166032). The change basically makes the `update_page` function work the same way as [the one you updated here](https://github.com/zulip/zulip/pull/19871/commits/852891413141423f4faaa8711c7be63c79f20981) for `realm_user_settings_defaults`, which is what I think you meant from your feedback.

If this is the change you had in mind, would it make sense to have both `user_settings` and `realm_default_user_settings` import and use the same function for the page update? And does it make sense to do this change without making the settings that require a reload of the page to be live updates as well?

**Testing plan:** I've done some manual testing, but would love feedback on whether writing a test to confirm the behavior makes sense. One thing that seems broken (when I test with or without these changes) is the emojiset update (see GIF below) as it doesn't update until the page does a hard refresh, but (unlike the user sidebar update and language update) no note is given to the user about that.

**GIFs or screenshots:**
### Updates to color scheme, user sidebar location and emojiset
![Zulip_fix_update_user_settings](https://user-images.githubusercontent.com/63245456/140799944-8e498acd-9289-416b-9bed-b3993285e821.gif)

### Update language (log out / log in)
![Zulip_language_update_user_settings](https://user-images.githubusercontent.com/63245456/140800811-19791cb5-8343-43e1-8b33-1dd699914f4a.gif)

timabbott commented on static/js/settings_display.js:
```diff
@@ -179,28 +179,18 @@ export async function report_emojiset_change(settings_panel) {
+    const value = user_settings[property];
 
     // TODO: Set emoji set selector here.
     // Longer term, we'll want to automate this function
```
I think this comment probably can be deleted.  Though it's worth looking at whether we can make `settings_org.set_input_element_value` handle the `emojiset` type of element as well, the "automate this function" part is clearly addressed by calling that function rather than doing each element by hand.

sahil839 commented on static/js/server_events_dispatch.js:
```diff
@@ -678,7 +678,7 @@ export function dispatch_normal_event(event) {
                 $("#go-to-default-view-hotkey-help").toggleClass("notdisplayed", !event.value);
                 break;
```
We need to remove this `break` statement to correctly update `Escape key navigates to default view` setting.

sahil839 reviewed (commented):
The 24-hour time setting is not updated correctly here and in user-default settings as well. I added the commit to fix it in #20204 which was for fixing another bug in default settings.

> If this is the change you had in mind, would it make sense to have both user_settings and realm_default_user_settings import and use the same function for the page update?

Yes we can use the same function for both of them.

I am not sure of the live-updating the setting that require reloads, but I think we can skip them (note that we do not update language setting as per this PR).

timabbott commented:
@laurynmm is this ready to be merged?  I see it still has the WIP tag.

laurynmm commented:
> @laurynmm is this ready to be merged? I see it still has the WIP tag.

@timabbott Well ... I'm not sure. I updated the commit to address your thought on having the `settings_org.set_input_value_element` handle the `emojiset` type. It seemed to be working, so I rebased and pushed my branch. It seemed like the `settings_org.get_input_value_element` was able to access the `emojiset` type, so I used that same logic in the other function.

But then I saw @sahil839 's comment / review and I was looking at his changes to the `realm_default_user_settings.update_page`. I'm trying to understand why the 24hr time needs to be changed from a boolean to a string.

sahil839 commented:
In this PR itself on which you are working. To be clear, I am talking about the language on the setting button.

sahil839 commented:
> But then I saw @sahil839 's comment / review and I was looking at his changes to the `realm_default_user_settings.update_page`. I'm trying to understand why the 24hr time needs to be changed from a boolean to a string.

The `user_settings[property]` for `twenty_four_hour_time` would be a boolean true/false, so in `settings_org.set_input_element_value` the input_type will be considered as boolean and the setting will be treated as checkbox which is not correct.

laurynmm commented:
@sahil839 - Yes, thank you for the clarification and nudge. I see what I was missing now re: both language and 24hr time. Getting on fixing both of those issues now.

laurynmm commented:
@sahil839 - I've updated my changes to reflect [your fix to the realm default settings](#20204) from yesterday with @timabbott's comments. Per this PR, there are two small differences between `settings_display.update_page` and `settings_realm_user_settings.update_page`.

1. `settings_display` updates the default language button text. The default language setting hasn't been implemented yet for the realm level defaults, so this isn't currently needed there.
2. `settings_realm_user_settings` has a check for the overlays at the start. I think this is because of the save/discard button implementation for these realm default settings. Since the user settings are implemented immediately on the change, it doesn't seem necessary to do verify the overlay is open. But I can add it if my understanding is incorrect.

sahil839 commented:
The check for overlay is to ensure that we do not unnecessarily update the elements since when overlay is opened, it will open with updated values only. There is no relation with save-discard widget.

laurynmm commented:
> The check for overlay is to ensure that we do not unnecessarily update the elements since when overlay is opened, it will open with updated values only. There is no relation with save-discard widget.

Ah, since it wasn't included in the pre-refactor `settings_display.update_page` function, I thought it was specific to the implementation for the realm settings. But really it should have been there so that the update wasn't being called unnecessarily.

sahil839 commented:
> 1. `settings_display` updates the default language button text. The default language setting hasn't been implemented yet for the realm level defaults, so this isn't currently needed there.

We will not have language setting for realm-level defaults since we would want to directly use browser's language (#16938). I think we can still deduplicate both the `update_page` functions in a follow-up commit/PR.

timabbott commented:
This is great, merged, thanks @laurynmm!  It's always great to see changes like this that reduce the complexity of the system.
